File: identidad/internal/shared/application/validacion_password.go
package application

import (
	"fmt"
	"unicode"
)

// ValidarFormatoPassword valida que una contraseña cumpla con los requisitos de seguridad:
// - Mínimo 8 caracteres
// - Al menos 1 mayúscula
// - Al menos 1 minúscula
// - Al menos 1 número
// - Al menos 1 carácter no alfanumérico
func ValidarFormatoPassword(password, campo string) error {
	if len(password) < 8 {
		return fmt.Errorf("%s debe tener al menos 8 caracteres", campo)
	}
	var mayuscula, minuscula, numero, noAlfanumerico bool
	for _, c := range password {
		switch {
		case unicode.IsUpper(c):
			mayuscula = true
		case unicode.IsLower(c):
			minuscula = true
		case unicode.IsDigit(c):
			numero = true
		case !unicode.IsLetter(c) && !unicode.IsDigit(c):
			noAlfanumerico = true
		}
	}
	if !mayuscula {
		return fmt.Errorf("%s debe contener al menos una mayúscula", campo)
	}
	if !minuscula {
		return fmt.Errorf("%s debe contener al menos una minúscula", campo)
	}
	if !numero {
		return fmt.Errorf("%s debe contener al menos un número", campo)
	}
	if !noAlfanumerico {
		return fmt.Errorf("%s debe contener al menos un carácter no alfanumérico", campo)
	}
	return nil
}

File: frontend/src/services/validacionPassword.js
/**
 * Valida formato de contraseña según reglas de seguridad.
 * @param {string} password
 * @returns {{ valida: boolean, errores: string[] }}
 */
export function validarPassword(password) {
  const errores = [];

  if (!password) {
    errores.push('La contraseña es requerida');
    return { valida: false, errores };
  }

  if (password.length < 8) {
    errores.push('Debe tener al menos 8 caracteres');
  }

  let mayuscula = false, minuscula = false, numero = false, noAlfanumerico = false;

  for (const c of password) {
    if (c >= 'A' && c <= 'Z') mayuscula = true;
    else if (c >= 'a' && c <= 'z') minuscula = true;
    else if (c >= '0' && c <= '9') numero = true;
    else if (!((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9'))) noAlfanumerico = true;
  }

  if (!mayuscula) errores.push('Debe contener al menos una mayúscula');
  if (!minuscula) errores.push('Debe contener al menos una minúscula');
  if (!numero) errores.push('Debe contener al menos un número');
  if (!noAlfanumerico) errores.push('Debe contener al menos un carácter especial (@, #, $, etc.)');

  return { valida: errores.length === 0, errores };
}
